File: market-data/crypto-news/examples/python/portfolio.py
#!/usr/bin/env python3
"""
Portfolio & Alerts API Examples - Python
Free Crypto News API - https://github.com/nirholas/free-crypto-news

Examples for portfolio management and alert endpoints.
"""

import requests
import json
from typing import Optional, List

BASE_URL = "https://cryptocurrency.cv"


# =============================================================================
# GET /api/portfolio - Get Portfolio
# =============================================================================

def get_portfolio(api_key: str) -> dict:
    """
    Get user's portfolio.
    
    Args:
        api_key: Your API key
    
    Returns:
        Portfolio holdings and value
    """
    headers = {"X-API-Key": api_key}
    response = requests.get(f"{BASE_URL}/api/portfolio", headers=headers)
    return response.json()


# =============================================================================
# POST /api/portfolio - Create/Update Portfolio
# =============================================================================

def create_portfolio(api_key: str, name: str, holdings: List[dict]) -> dict:
    """
    Create or update portfolio.
    
    Args:
        api_key: Your API key
        name: Portfolio name
        holdings: List of holdings [{symbol, amount, price}]
    
    Returns:
        Created portfolio
    """
    headers = {
        "X-API-Key": api_key,
        "Content-Type": "application/json"
    }
    payload = {
        "name": name,
        "holdings": holdings
    }
    response = requests.post(f"{BASE_URL}/api/portfolio", 
                            headers=headers, json=payload)
    return response.json()


# =============================================================================
# POST /api/portfolio/add - Add Holding
# =============================================================================

def add_holding(api_key: str, symbol: str, amount: float, price: float) -> dict:
    """
    Add a holding to portfolio.
    
    Args:
        api_key: Your API key
        symbol: Asset symbol (BTC, ETH, etc.)
        amount: Amount of asset
        price: Purchase price
    
    Returns:
        Updated portfolio
    """
    headers = {
        "X-API-Key": api_key,
        "Content-Type": "application/json"
    }
    payload = {
        "symbol": symbol,
        "amount": amount,
        "price": price
    }
    response = requests.post(f"{BASE_URL}/api/portfolio/add",
                            headers=headers, json=payload)
    return response.json()


# =============================================================================
# GET /api/portfolio/performance - Portfolio Performance
# =============================================================================

def get_portfolio_performance(api_key: str, period: str = "30d") -> dict:
    """
    Get portfolio performance over time.
    
    Args:
        api_key: Your API key
        period: Time period (7d, 30d, 90d, 1y)
    
    Returns:
        Performance metrics
    """
    headers = {"X-API-Key": api_key}
    params = {"period": period}
    response = requests.get(f"{BASE_URL}/api/portfolio/performance",
                           headers=headers, params=params)
    return response.json()


# =============================================================================
# GET /api/portfolio/news - Portfolio-Related News
# =============================================================================

def get_portfolio_news(api_key: str, limit: int = 20) -> dict:
    """
    Get news related to portfolio holdings.
    
    Args:
        api_key: Your API key
        limit: Number of articles
    
    Returns:
        News for portfolio assets
    """
    headers = {"X-API-Key": api_key}
    params = {"limit": limit}
    response = requests.get(f"{BASE_URL}/api/portfolio/news",
                           headers=headers, params=params)
    return response.json()


# =============================================================================
# GET /api/alerts - Get Alerts
# =============================================================================

def get_alerts(api_key: str) -> dict:
    """
    Get all configured alerts.
    
    Args:
        api_key: Your API key
    
    Returns:
        List of alerts
    """
    headers = {"X-API-Key": api_key}
    response = requests.get(f"{BASE_URL}/api/alerts", headers=headers)
    return response.json()


# =============================================================================
# POST /api/alerts - Create Alert
# =============================================================================

def create_alert(api_key: str, type: str, asset: str, 
                 condition: str, value: float,
                 notification: str = "email") -> dict:
    """
    Create a new alert.
    
    Args:
        api_key: Your API key
        type: Alert type (price, sentiment, volume, news)
        asset: Asset symbol
        condition: Condition (above, below, change)
        value: Trigger value
        notification: Notification method (email, webhook, push)
    
    Returns:
        Created alert
    """
    headers = {
        "X-API-Key": api_key,
        "Content-Type": "application/json"
    }
    payload = {
        "type": type,
        "asset": asset,
        "condition": condition,
        "value": value,
        "notification": notification
    }
    response = requests.post(f"{BASE_URL}/api/alerts",
                            headers=headers, json=payload)
    return response.json()


# =============================================================================
# DELETE /api/alerts/[id] - Delete Alert
# =============================================================================

def delete_alert(api_key: str, alert_id: str) -> dict:
    """
    Delete an alert.
    
    Args:
        api_key: Your API key
        alert_id: Alert ID
    
    Returns:
        Deletion result
    """
    headers = {"X-API-Key": api_key}
    response = requests.delete(f"{BASE_URL}/api/alerts/{alert_id}",
                              headers=headers)
    return response.json()


# =============================================================================
# GET /api/watchlist - Get Watchlist
# =============================================================================

def get_watchlist(api_key: str) -> dict:
    """
    Get user's watchlist.
    
    Args:
        api_key: Your API key
    
    Returns:
        Watchlist assets
    """
    headers = {"X-API-Key": api_key}
    response = requests.get(f"{BASE_URL}/api/watchlist", headers=headers)
    return response.json()


# =============================================================================
# POST /api/watchlist - Add to Watchlist
# =============================================================================

def add_to_watchlist(api_key: str, symbol: str) -> dict:
    """
    Add asset to watchlist.
    
    Args:
        api_key: Your API key
        symbol: Asset symbol
    
    Returns:
        Updated watchlist
    """
    headers = {
        "X-API-Key": api_key,
        "Content-Type": "application/json"
    }
    response = requests.post(f"{BASE_URL}/api/watchlist",
                            headers=headers, json={"symbol": symbol})
    return response.json()


# =============================================================================
# DELETE /api/watchlist/[symbol] - Remove from Watchlist
# =============================================================================

def remove_from_watchlist(api_key: str, symbol: str) -> dict:
    """
    Remove asset from watchlist.
    
    Args:
        api_key: Your API key
        symbol: Asset symbol
    
    Returns:
        Updated watchlist
    """
    headers = {"X-API-Key": api_key}
    response = requests.delete(f"{BASE_URL}/api/watchlist/{symbol}",
                              headers=headers)
    return response.json()


# =============================================================================
# GET /api/notifications - Get Notifications
# =============================================================================

def get_notifications(api_key: str, unread_only: bool = False) -> dict:
    """
    Get user notifications.
    
    Args:
        api_key: Your API key
        unread_only: Only unread notifications
    
    Returns:
        Notifications list
    """
    headers = {"X-API-Key": api_key}
    params = {"unread": str(unread_only).lower()}
    response = requests.get(f"{BASE_URL}/api/notifications",
                           headers=headers, params=params)
    return response.json()


# =============================================================================
# POST /api/notifications/read - Mark as Read
# =============================================================================

def mark_notifications_read(api_key: str, notification_ids: List[str]) -> dict:
    """
    Mark notifications as read.
    
    Args:
        api_key: Your API key
        notification_ids: List of notification IDs
    
    Returns:
        Result
    """
    headers = {
        "X-API-Key": api_key,
        "Content-Type": "application/json"
    }
    response = requests.post(f"{BASE_URL}/api/notifications/read",
                            headers=headers, json={"ids": notification_ids})
    return response.json()


# =============================================================================
# GET /api/preferences - Get Preferences
# =============================================================================

def get_preferences(api_key: str) -> dict:
    """
    Get user preferences.
    
    Args:
        api_key: Your API key
    
    Returns:
        User preferences
    """
    headers = {"X-API-Key": api_key}
    response = requests.get(f"{BASE_URL}/api/preferences", headers=headers)
    return response.json()


# =============================================================================
# PUT /api/preferences - Update Preferences
# =============================================================================

def update_preferences(api_key: str, preferences: dict) -> dict:
    """
    Update user preferences.
    
    Args:
        api_key: Your API key
        preferences: Preference settings
    
    Returns:
        Updated preferences
    """
    headers = {
        "X-API-Key": api_key,
        "Content-Type": "application/json"
    }
    response = requests.put(f"{BASE_URL}/api/preferences",
                           headers=headers, json=preferences)
    return response.json()


# =============================================================================
# COMPLETE EXAMPLES
# =============================================================================

if __name__ == "__main__":
    print("\n" + "="*60)
    print("FREE CRYPTO NEWS API - PORTFOLIO & ALERTS EXAMPLES")
    print("="*60)
    
    # Note: These require an API key
    API_KEY = "your-api-key-here"
    
    print("\n⚠️ These endpoints require authentication.")
    print("   Replace 'your-api-key-here' with your actual API key.")
    
    # Example 1: Create Portfolio
    print("\n📊 1. Create Portfolio")
    portfolio = create_portfolio(
        api_key=API_KEY,
        name="My Crypto Portfolio",
        holdings=[
            {"symbol": "BTC", "amount": 0.5, "price": 45000},
            {"symbol": "ETH", "amount": 5.0, "price": 2500},
            {"symbol": "SOL", "amount": 50, "price": 100}
        ]
    )
    print(f"   Portfolio: {portfolio}")
    
    # Example 2: Create Price Alert
    print("\n🔔 2. Create Price Alert")
    alert = create_alert(
        api_key=API_KEY,
        type="price",
        asset="BTC",
        condition="above",
        value=100000,
        notification="email"
    )
    print(f"   Alert: {alert}")
    
    # Example 3: Create Sentiment Alert
    print("\n📈 3. Create Sentiment Alert")
    sentiment_alert = create_alert(
        api_key=API_KEY,
        type="sentiment",
        asset="ETH",
        condition="below",
        value=0.3,  # Sentiment drops below 0.3
        notification="webhook"
    )
    print(f"   Sentiment Alert: {sentiment_alert}")
    
    # Example 4: Add to Watchlist
    print("\n👀 4. Add to Watchlist")
    watchlist = add_to_watchlist(API_KEY, "AVAX")
    print(f"   Watchlist: {watchlist}")
    
    # Example 5: Get Portfolio News
    print("\n📰 5. Get Portfolio News")
    news = get_portfolio_news(API_KEY, limit=5)
    print(f"   News: {news}")
    
    # Example 6: Update Preferences
    print("\n⚙️ 6. Update Preferences")
    prefs = update_preferences(API_KEY, {
        "language": "en",
        "timezone": "UTC",
        "notifications": {
            "email": True,
            "push": True,
            "digest": "daily"
        },
        "categories": ["bitcoin", "ethereum", "defi"]
    })
    print(f"   Preferences: {prefs}")
    
    print("\n" + "="*60)
    print("Portfolio & alerts examples completed!")
    print("="*60)

File: market-data/crypto-news/examples/javascript/portfolio.js
/**
 * Free Crypto News API - Portfolio & User Examples
 * https://github.com/nirholas/free-crypto-news
 * 
 * Portfolio tracking, alerts, watchlists, and user preferences.
 * Note: These endpoints require authentication (API key).
 */

const BASE_URL = 'https://cryptocurrency.cv';

// =============================================================================
// HELPER
// =============================================================================

/**
 * Create authenticated fetch helper
 * @param {string} apiKey - Your API key
 * @returns {Function} Authenticated fetch function
 */
function createAuthFetch(apiKey) {
  return async (url, options = {}) => {
    const response = await fetch(url, {
      ...options,
      headers: {
        ...options.headers,
        'X-API-Key': apiKey,
        'Content-Type': 'application/json',
      },
    });
    return response.json();
  };
}

// =============================================================================
// PORTFOLIO
// =============================================================================

/**
 * Get portfolio
 * @param {string} apiKey - API key
 * @returns {Promise<Object>} Portfolio data
 */
async function getPortfolio(apiKey) {
  const authFetch = createAuthFetch(apiKey);
  return authFetch(`${BASE_URL}/api/portfolio`);
}

/**
 * Create or update portfolio
 * @param {string} apiKey - API key
 * @param {Object} portfolio - Portfolio data
 * @returns {Promise<Object>} Updated portfolio
 */
async function updatePortfolio(apiKey, portfolio) {
  const authFetch = createAuthFetch(apiKey);
  return authFetch(`${BASE_URL}/api/portfolio`, {
    method: 'POST',
    body: JSON.stringify(portfolio),
  });
}

/**
 * Add holding to portfolio
 * @param {string} apiKey - API key
 * @param {Object} holding - Holding data
 * @returns {Promise<Object>} Updated portfolio
 */
async function addHolding(apiKey, holding) {
  const authFetch = createAuthFetch(apiKey);
  return authFetch(`${BASE_URL}/api/portfolio/holdings`, {
    method: 'POST',
    body: JSON.stringify(holding),
  });
}

/**
 * Remove holding from portfolio
 * @param {string} apiKey - API key
 * @param {string} holdingId - Holding ID
 * @returns {Promise<Object>} Updated portfolio
 */
async function removeHolding(apiKey, holdingId) {
  const authFetch = createAuthFetch(apiKey);
  return authFetch(`${BASE_URL}/api/portfolio/holdings/${holdingId}`, {
    method: 'DELETE',
  });
}

/**
 * Get portfolio performance
 * @param {string} apiKey - API key
 * @param {Object} options - Query options
 * @returns {Promise<Object>} Performance data
 */
async function getPortfolioPerformance(apiKey, options = {}) {
  const params = new URLSearchParams({
    period: options.period || '30d',
    ...options
  });
  const authFetch = createAuthFetch(apiKey);
  return authFetch(`${BASE_URL}/api/portfolio/performance?${params}`);
}

/**
 * Get news related to portfolio holdings
 * @param {string} apiKey - API key
 * @param {Object} options - Query options
 * @returns {Promise<Object>} Portfolio news
 */
async function getPortfolioNews(apiKey, options = {}) {
  const params = new URLSearchParams({
    limit: options.limit || 20,
    ...options
  });
  const authFetch = createAuthFetch(apiKey);
  return authFetch(`${BASE_URL}/api/portfolio/news?${params}`);
}

// =============================================================================
// ALERTS
// =============================================================================

/**
 * Get user alerts
 * @param {string} apiKey - API key
 * @returns {Promise<Object>} Alerts list
 */
async function getAlerts(apiKey) {
  const authFetch = createAuthFetch(apiKey);
  return authFetch(`${BASE_URL}/api/alerts`);
}

/**
 * Create alert
 * @param {string} apiKey - API key
 * @param {Object} alert - Alert configuration
 * @returns {Promise<Object>} Created alert
 */
async function createAlert(apiKey, alert) {
  const authFetch = createAuthFetch(apiKey);
  return authFetch(`${BASE_URL}/api/alerts`, {
    method: 'POST',
    body: JSON.stringify(alert),
  });
}

/**
 * Update alert
 * @param {string} apiKey - API key
 * @param {string} alertId - Alert ID
 * @param {Object} updates - Alert updates
 * @returns {Promise<Object>} Updated alert
 */
async function updateAlert(apiKey, alertId, updates) {
  const authFetch = createAuthFetch(apiKey);
  return authFetch(`${BASE_URL}/api/alerts/${alertId}`, {
    method: 'PATCH',
    body: JSON.stringify(updates),
  });
}

/**
 * Delete alert
 * @param {string} apiKey - API key
 * @param {string} alertId - Alert ID
 * @returns {Promise<Object>} Delete response
 */
async function deleteAlert(apiKey, alertId) {
  const authFetch = createAuthFetch(apiKey);
  return authFetch(`${BASE_URL}/api/alerts/${alertId}`, {
    method: 'DELETE',
  });
}

// =============================================================================
// WATCHLIST
// =============================================================================

/**
 * Get watchlist
 * @param {string} apiKey - API key
 * @returns {Promise<Object>} Watchlist
 */
async function getWatchlist(apiKey) {
  const authFetch = createAuthFetch(apiKey);
  return authFetch(`${BASE_URL}/api/watchlist`);
}

/**
 * Add to watchlist
 * @param {string} apiKey - API key
 * @param {string} symbol - Asset symbol
 * @returns {Promise<Object>} Updated watchlist
 */
async function addToWatchlist(apiKey, symbol) {
  const authFetch = createAuthFetch(apiKey);
  return authFetch(`${BASE_URL}/api/watchlist`, {
    method: 'POST',
    body: JSON.stringify({ symbol }),
  });
}

/**
 * Remove from watchlist
 * @param {string} apiKey - API key
 * @param {string} symbol - Asset symbol
 * @returns {Promise<Object>} Updated watchlist
 */
async function removeFromWatchlist(apiKey, symbol) {
  const authFetch = createAuthFetch(apiKey);
  return authFetch(`${BASE_URL}/api/watchlist/${symbol}`, {
    method: 'DELETE',
  });
}

// =============================================================================
// NOTIFICATIONS
// =============================================================================

/**
 * Get notifications
 * @param {string} apiKey - API key
 * @param {Object} options - Query options
 * @returns {Promise<Object>} Notifications
 */
async function getNotifications(apiKey, options = {}) {
  const params = new URLSearchParams({
    unread: options.unread || false,
    limit: options.limit || 50,
    ...options
  });
  const authFetch = createAuthFetch(apiKey);
  return authFetch(`${BASE_URL}/api/notifications?${params}`);
}

/**
 * Mark notifications as read
 * @param {string} apiKey - API key
 * @param {string[]} ids - Notification IDs
 * @returns {Promise<Object>} Update response
 */
async function markNotificationsRead(apiKey, ids) {
  const authFetch = createAuthFetch(apiKey);
  return authFetch(`${BASE_URL}/api/notifications/read`, {
    method: 'POST',
    body: JSON.stringify({ ids }),
  });
}

/**
 * Mark all notifications as read
 * @param {string} apiKey - API key
 * @returns {Promise<Object>} Update response
 */
async function markAllRead(apiKey) {
  const authFetch = createAuthFetch(apiKey);
  return authFetch(`${BASE_URL}/api/notifications/read-all`, {
    method: 'POST',
  });
}

// =============================================================================
// PREFERENCES
// =============================================================================

/**
 * Get user preferences
 * @param {string} apiKey - API key
 * @returns {Promise<Object>} User preferences
 */
async function getPreferences(apiKey) {
  const authFetch = createAuthFetch(apiKey);
  return authFetch(`${BASE_URL}/api/preferences`);
}

/**
 * Update preferences
 * @param {string} apiKey - API key
 * @param {Object} preferences - Preference updates
 * @returns {Promise<Object>} Updated preferences
 */
async function updatePreferences(apiKey, preferences) {
  const authFetch = createAuthFetch(apiKey);
  return authFetch(`${BASE_URL}/api/preferences`, {
    method: 'PATCH',
    body: JSON.stringify(preferences),
  });
}

// =============================================================================
// EXPORTS
// =============================================================================

module.exports = {
  // Helper
  createAuthFetch,
  
  // Portfolio
  getPortfolio,
  updatePortfolio,
  addHolding,
  removeHolding,
  getPortfolioPerformance,
  getPortfolioNews,
  
  // Alerts
  getAlerts,
  createAlert,
  updateAlert,
  deleteAlert,
  
  // Watchlist
  getWatchlist,
  addToWatchlist,
  removeFromWatchlist,
  
  // Notifications
  getNotifications,
  markNotificationsRead,
  markAllRead,
  
  // Preferences
  getPreferences,
  updatePreferences,
};

// =============================================================================
// EXAMPLE USAGE
// =============================================================================

async function runExamples() {
  console.log('\n=== Portfolio & User API Examples ===\n');
  console.log('Note: These examples require a valid API key.\n');
  
  const API_KEY = process.env.CRYPTO_NEWS_API_KEY || 'your-api-key';
  
  if (API_KEY === 'your-api-key') {
    console.log('⚠️  Set CRYPTO_NEWS_API_KEY environment variable to run examples.');
    console.log('   export CRYPTO_NEWS_API_KEY="your-actual-api-key"');
    return;
  }
  
  try {
    // Portfolio
    console.log('1. Fetching portfolio...');
    const portfolio = await getPortfolio(API_KEY);
    console.log(`   Holdings: ${portfolio.holdings?.length || 0}`);
    
    // Add holding
    console.log('2. Adding BTC to portfolio...');
    const added = await addHolding(API_KEY, {
      symbol: 'BTC',
      amount: 0.5,
      purchasePrice: 40000,
      purchaseDate: '2024-01-15',
    });
    console.log(`   Result: ${added.success ? 'Success' : 'Failed'}`);
    
    // Performance
    console.log('3. Fetching portfolio performance...');
    const performance = await getPortfolioPerformance(API_KEY, { period: '30d' });
    console.log(`   Return: ${performance.totalReturn || 'N/A'}%`);
    
    // Alerts
    console.log('4. Fetching alerts...');
    const alerts = await getAlerts(API_KEY);
    console.log(`   Active alerts: ${alerts.data?.length || 0}`);
    
    // Create alert
    console.log('5. Creating price alert...');
    const alert = await createAlert(API_KEY, {
      type: 'price',
      asset: 'BTC',
      condition: 'above',
      value: 100000,
      channels: ['email', 'push'],
    });
    console.log(`   Alert created: ${alert.id || 'N/A'}`);
    
    // Watchlist
    console.log('6. Fetching watchlist...');
    const watchlist = await getWatchlist(API_KEY);
    console.log(`   Watching: ${watchlist.assets?.length || 0} assets`);
    
    // Notifications
    console.log('7. Fetching notifications...');
    const notifications = await getNotifications(API_KEY, { unread: true });
    console.log(`   Unread: ${notifications.data?.length || 0}`);
    
    // Preferences
    console.log('8. Fetching preferences...');
    const prefs = await getPreferences(API_KEY);
    console.log(`   Theme: ${prefs.theme || 'default'}`);
    
    console.log('\n✅ All portfolio examples completed!\n');
  } catch (error) {
    console.error('Error:', error.message);
  }
}

// Run if executed directly
if (require.main === module) {
  runExamples();
}
